feat: add npm audit command

jdanil opened the pull request:
**What's the problem this PR addresses?**

Implement `yarn audit` command from yarn classic.

Closes #1187
Closes #1461

**How did you fix it?**

Added a new `audit` command to the existing `npm-cli` plugin.

Added a new npm HTTP `post` util (clone of `put`).

Code to get transitive dev dependencies is heavily adapted from [yarn classic](https://github.com/yarnpkg/yarn/blob/master/src/util/get-transitive-dev-dependencies.js).

**Checklist**

<!--- Don't worry if you miss something, chores are automatically tested. -->
<!--- This checklist exists to help you remember doing the chores when you submit a PR. -->
<!--- Put an `x` in all the boxes that apply. -->
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

<!-- See https://yarnpkg.com/advanced/contributing#preparing-your-pr-to-be-released for more details. -->
<!-- Check with `yarn version check` and fix with `yarn version check -i` -->
- [x] I have set the packages that need to be released for my changes to be effective.

<!-- The "Testing chores" workflow validates that your PR follows our guidelines. -->
<!-- If it doesn't pass, click on it to see details as to what your PR might be missing. -->
- [ ] I will check that all automated PR checks pass before the PR gets reviewed.

jdanil commented on packages/plugin-npm-cli/sources/commands/npm/auditUtils.ts:
```diff
@@ -0,0 +1,69 @@
+
+  while (queue.length > 0) {
+    const descriptorHash = queue.shift();
+    const locatorHash = storedResolutions.get(descriptorHash as DescriptorHash);
```
I'm having a bit of trouble here. For some reason none of the descriptor hashes passed in from `roots` have had entries in `storedResolutions`. It keeps returning `undefined` so there is no `originalPackage` found, and the rest of the code in this loop is skipped.

arcanis replied:
Usually this is caused when you forget to run `await project.restoreInstallState()` at the beginning of your Command implementation (cf `run.ts` for an example).

jdanil replied:
Ohh I certainly didn't do this. Thanks for the tip!

jdanil replied:
I got a bit closer on this issue today.

I found that when I try to use a `descriptorHash` from `workspace.manifest.dependencies.values()` to get a `locatorHash` from `project.storedResolutions`, it always returns `undefined`.
But if I use `workspace.dependencies.values()` instead, it is able to retrieve the locator hashes.

`workspace.dependencies` includes dependencies + devDependencies + workspaces. And for this part I need to be able to distinguish dependencies from devDependencies, so I can't really use it without some other way of identifying the dependency type.

If anyone has any ideas why this is happening, let me know. I'll keep plugging away in the meantime.

arcanis replied:
Descriptors from `manifest.dependencies` are the raw descriptors, exactly as listed in the `package.json` file. Because we normalize them before resolving them (`foo@1.2.3` -> `foo@npm:1.2.3`), they probably aren't part of the `storedResolution` map.

> And for this part I need to be able to distinguish dependencies from devDependencies, so I can't really use it without some other way of identifying the dependency type.

You probably can do a bit of both: use the `workspace.dependencies` set, then check whether the dependencies' ident is part of `workspace.manifest.devDependencies`.

jdanil replied:
Ahh right. I didn't realise there was a difference as they used the same types.
Thanks for the pointer! This is working now.

jdanil commented on packages/plugin-npm-cli/sources/commands/npm/audit.ts:
I haven't added any pretty outputs yet. Right now if there is an error it returns the whole JSON report, and if it passes it exits without printing.
Should I simply add output here to match existing behaviour in yarn classic?

Also wasn't sure if I needed to create a new exception code, or if I can reuse the generic `EXCEPTION`.

If there are multiple vulnerabilities, should they each be their own individual error in the report, or a single error?

arcanis replied:
We'll need the default display to be pretty, and the JSON stream to be returned with the `--json` flag (to be consistent with the other commands). We don't need to (and we probably shouldn't) follow the output from Yarn Classic - I prefer to careful craft an appealing output without being burdened by the past decisions.

In this particular case, I think the best would be to use the new `emitTree` function from `treeUtils` in order to print the tree of 'vulnerable packages' (with only the associated CVEs), and perhaps to provide more details when the `-v,--verbose` flag is set.

jdanil replied:
Cheers @arcanis. I did implement the `--json` flag and used the `reportJson` function just above this, so that should all be working. Just didn't want to touch the human readable output until I had an idea what you guys were after.

Will checkout the `emitTree` function. Thanks for the pointer 👍

jdanil commented on .yarn/versions/61ef8dac.yml:
```diff
@@ -0,0 +1,3 @@
+releases:
+  "@yarnpkg/plugin-npm": minor
+  "@yarnpkg/plugin-npm-cli": minor
```
I used `yarn version check --interactive` here.

These are the two packages I touched. I selected `unsure` for all the other packages that depend on these.
Should they be patch/minor releases? If not, do I need to explicitly list all of them here as `declined`?

arcanis replied:
Minor for those two is correct! I think the only other required is `@yarnpkg/cli` which should be "minor" as well. For the rest, select "declined" as we won't need to release them for this feature.

jdanil replied:
Will update. Thanks 🙏

arcanis commented:
cc @lirantal - is there something we can do to be compatible with Snyk as well? Any improvements or suggestion?

lirantal commented:
Thanks for the ping @arcanis and really great to see this taking shape @jdanil 👏 

Things to note and consider:
1. Ignoring devDependencies by default - we're doing this with snyk and the snyk CLI and it greatly reduces the signal to noise ratio, which developers often complain about (as in, regular expression denial of service in jest or webpack)
2. Instead of reporting vulnerabilities as all the paths which resolve to a specific and single CVE report it as one, and if needed we can print the paths. npm prints all the paths instead of just showing the one vulnerability which results in some silly high numbers like `found 13,391` where-as there are a total of about 2000 CVEs all in all in the npm ecosystem, so that 13k number just doesn't make sense. Snyk here is also aligned with reporting just one CVE instead of counting all vulnerable paths.
3. I find the report output of npm audit hard to follow. It's more of a "report" rather than driving you to action as a developer to solve the problem. Here's out an example `snyk test` output looks like for a project I'm working on:

![image](https://user-images.githubusercontent.com/316371/94661152-5e837b80-030f-11eb-939c-18fa2b716765.png)


You can notice a few things:
* It doesn't need to iterate every single vulnerable path as there could be thousands and it doesn't scale to track each path.
* Instead of per-module reporting, it actually groups the results per the necessary upgrade and what it fixes. Driving a clear action that is needed from a developer: "Ok I'll upgrade express to 4.16.0 and fix all of those 6 vulnerabilities"

4. Lastly - I recommend we use the Snyk API as it is meant to be used for this instead of having to reverse engineer the npm audit endpoint. I talked about this with @arcanis in the past and we were happy to make the change.

Great stuff and happy to see this coming through.

jdanil commented:
Hi @lirantal 👋 Thanks for putting together those notes.

1. I just pushed a change to support filtering by environment (i.e dependencies and devDependencies). I'm on the fence about whether devDependencies should be ignored by default. I've had the same thought in the past, but after reading @quinnturner's [comment](https://github.com/IBM/audit-ci/issues/91#issuecomment-499550911) on audit-ci I'm thinking auditing all dependencies by default might be the safer approach. Have you found any issues with this approach in snyk? I've set the default to "all" for now, but happy to update it once there is consensus.
1. This approach sounds good. I'll try implementing this soon.
1. This is really neat. It sounds like there are two different outputs for snyk? If I'm understanding this correctly, in your 2nd point it sounds like the output is grouped by CVE, but in your 3rd point it sounds like the output is grouped by direct dependencies of the project. Are you suggesting that we support both outputs?
1. I had a quick look at the snyk API. Is there a free version for public use? I haven't been able to track one down, and I noticed this in the docs "The Snyk API is available to customers on paid plans and allows you to programatically integrate with Snyk". I'm also interested in what the difference is between the npm and snyk auditing functionality. Is it just that they each maintain their own collection of known vulnerabilities? Is the output the same?

Just to update you guys on where I'm at...

- @arcanis I haven't been able to find anything on a new beta API endpoint from npm. If you have any more information on that I can have a look again, otherwise maybe we just use the existing API and update the implementation if/when a new one is made available.
- I added a `--recursive` flag, so that by default only dependencies in the current workspace are audited. But when the flag is passed as true, it will audit the whole project. This behaviour seemed to be more in like with the `info` and `why` commands. The implementation isn't really recursive though. If its true, its not like it only audits child workspaces, it will audit all workspaces in the project. Not sure if it should be update to limit it to that workspace's children (I don't think nested workspaces are a thing in yarn?), or if I rename the flag from "recursive" to "all", or if the current behaviour is okay.
- If we decide to move to the snyk API or npm beta API (if it exists), then I'll need to update this.
- Implement the user-friendly output once we agree on the approach. JSON is currently implemented (just outputs the API response).
- I'll start looking at the tests last, once the output is confirmed and implemented.
- I'm not too sure what to do about the "testing chores" verification job. It seems to be reporting the snake_case keys as issues, but these keys correspond to the API request/response format. Is there something I need to do to tell the linter to ignore them?

lirantal commented:
@jdanil see below my comments on each of your questions:

--
> Have you found any issues with this approach in snyk? 

Not at all and that's for a 1.5m developers using Snyk.
The thing is, the issue that @quinnturner is bringing up about malicious devDep is correct but if you will find out about it from an official vulnerability (i.e: a cve) that's going to be way too late in the game anyway :-) 

--

> This is really neat. It sounds like there are two different outputs for snyk? 

Nope, there's just one output :-)
I honestly suggest installing snyk (`npm install -g snyk`) and trying it out locally. I'm sure you'll find a lot of the developer experience there interesting and it'll be easier for you to make the analogies.

--

> I had a quick look at the snyk API. Is there a free version for public use? 

I'll be working on a new vulnerability-based API that will be free for anonymous use (so that yarn clients, which are end users) can ping it). It'll receive a list of packages and their versions (flat list), and will respond back with a list of vulnerabilities.
Something like:
API Input: 

```
{
 "marked": "2.3.2",
 "react: "0.15.0"
}
```

API Output:
```
[
  {
   "library": "marked",
    "version": "2.3.2",
    "cve": "2019-xxxx-yyyy",
    "severity": "high"
  }
]
```

Just a simple mock example but we can refine on the needed fields and information to have in the API.
It's plain and simple. No need to send a package-lock.json format or anything else. On the yarn side, pnpm side, etc the client-side logic can then turn the list of vulnerabilities into a dep tree to show the parent dependency introducing them etc.

What do you think?

If we're happy with moving forward on this, and with @arcanis's blessing, I can put up a mock API for you to send requests to while I build the vulnerability database up to date on it.

arcanis commented:
The main problem I see is that, as bad as it is, the reference npm audit protocol is the one used by the npm registry. So if we call the command `yarn npm audit`, it should be compatible with the npm protocol (and, presumably, use the npm dataset by default).

If we were to use a different, lighter protocol, it would necessarily have to be called something different (`yarn snyk audit`?), and should probably also live in an external `snyk` plugin (I'd prefer to avoid shipping proprietary integrations in the main bundle for now; I already really dislike adding the npm integrations, but at least people kinda expect it by now 😒).

I think for now the easiest is to keep going with the npm audit protocol since @jdanil nicely implemented it, follow the UX suggestion @lirantal's mentioned, and discuss about the Snyk protocol at a later time.

lirantal commented:
Honestly since npm has been a long-time for-profit company, it's odd not to think of it as a proprietary integration either. Now is even more the case, and also AFAIK there aren't any public API docs for the audit (or other) endpoints so that's where I feel it really isn't any less proprietary :-) 

I'm trying to do exactly the opposite here with Snyk - providing a documented, technology-agnostic and open endpoint for vulnerabilities for the ecosystem that can serve yarn, pnpm and others if they wanted. Having it as its own plugin for yarn is nice but considering that users would might as well just install the snyk CLI doesn't make a lot of sense to have it as a plugin.

If there's anything we learned about security and developers recently is that security on-by-default just goes a long way and the decisions made with the Snyk CLI that I pointed above address a lot of these. There are other reasons to use this like the rich vulnerability database that is far from what npm has. Recent two examples from frontend projects I worked on: Snyk has an [XSS vulnerability in vuetify](https://snyk.io/vuln/SNYK-JS-VUETIFY-474604) back from 2019 which npm audit doesn't report on, and in 2019 when I compared angular/react I found that Snyk reports on [11 AngularJS vulnerabilities which npm misses all of them](https://snyk.io/blog/angular-vs-react-security-bakeoff-2019).

Realizing how this might looked biased but I'm noting out actual technical and data-based differences that would end up benefiting end-users and it's genuinely my aim to help the community and I see no reason Yarn shouldn't have these upsides. I totally understand though if there's a decision otherwise to keep things the way they always were.

jdanil commented:
OK so I've just added some user-friendly output using the `emitTree` function. I've grouped it by advisory, and included each vulnerable path as tree nodes.
The alternative approach discussed was to group the vulnerabilities by direct dependencies, but I felt that would be a bit less useful as we are unable to provide recommendations on that level (i.e. we don't know what version a direct dependency needs to be upgraded to to depend on a patched version of the problematic dependency.

I also checked out npm@7 as they revamped their audit output too. I've confirmed that they're using the same version of the API (see https://www.npmjs.com/package/npm-audit-report#break-from-version-1). Our outputs are quite similar now (they're also grouping by advisory).

arcanis commented:
@jdanil Nice! I'm trying to run the command from the Berry repo but I consistently get `503: Service Unavailable`. Running `npm audit` in an npm project seems to work, so I suspect there's a payload error somewhere - does that ring a bell?

jdanil commented:
Ah I had been using another package to test the auditing, getting the yarn path to point to my local version of berry.
I hadn't actually run `yarn npm audit` on the berry repo itself 😅 

The issue was when sending the list of direct dependencies, it included the local workspace packages and the protocol (`workspace:`) was included in the version strings.
I've updated the PR to filter them out and use just the requires selector.

arcanis commented:
I've finally finished reviewing the code, and it's really nice - thanks for taking this up, @jdanil! I've done a few refactoring, mostly to bring the code in line with the code patterns used in the rest of the codebase (for instance I removed the `forEach` and `reduce` to replace them with classic `for ... of`, etc). I've also made a few logic changes, feel free to comment on them:

- I've removed the "Paths" section from the output. Give its size, it caused my terminal back buffer to overwrite, so I could only the the last two reports 😅 I think we should keep the critical information in line of sight, and people that want to know more can just run `yarn why` to know why the affected package is in the tree.

- I made it so that only reports for direct dependencies are printed by default. This one is perhaps a bit surprising given how `npm audit` works, but it brings the command behaviour in line with how most other commands of this kind work (by default operate on the direct dep, can be tweaked with `-A` and `-R`).

One question: do you happen to know what's the purpose of the `dev` fields from the `dependencies` payload? What happen if we were to set it to, say, an hardcoded `false`? I don't see how they could impact the reports.

jdanil commented:
No worries 🙂 Thanks for cleaning up the PR! Its much easier to parse after you've pulled out the types and utils into their own files.

Yeah after running audit on the berry repo for the first time, the "paths" section was a bit overwhelming. I think there was a lodash vulnerability so it was reporting in 100s of packages. ~I just tried to pull the branch so I could see what the updated behaviour looks like, but I'm getting a `400 (Bad Request)` error running `yarn npm audit` now. I'll have a deeper look into it a bit later today in case its something I'm doing wrong.~ (I was getting this error because I was running `yarn npm audit` without the `--recursive` flag. Will have a look later today if there is something that can be done to avoid this error.) But I was just going to say having the whole paths tree might be too much, but what do you think about just keeping the top-level (direct dependencies) in the report? That way users may be able to try updating those direct dependencies before investigating other strategies to resolve it like adding a resolution or investigating further with `yarn why`. The reason I think it might still be useful is that it is possible for a project to be using two version of a package in different places. This report will list just the ones using the vulnerable version whereas `yarn why` will list any uses or the dependency regardless of version.

Auditing only direct dependencies is interesting! I hadn't considered that. If I'm understanding the change correctly, `--all` will check dependencies for all workspaces in the project. And `--recursive` will include checks for transitive dependencies?
So the equivalent of `npm audit` would be `yarn npm audit --recursive` (and yarn classic equivalent of `yarn audit` would be `yarn npm audit --all --recursive`) but now we have the added functionality of being able to limit that audit report to just direct dependencies?

I don't believe that the `dev` field in the payload actually impacts what vulnerabilities are reported or their severity. From what I can gather, that field it just used to determine the "total" counts. If you run the command with the `--json` flag at the end of the output you will see that it includes a count of the number of dependencies, devDependencies, optionalDependencies, and totalDependencies.

jdanil commented:
Also, what are your thoughts on having some sort of summary?
I think in my last commit I had a bit of a half-baked idea of just printing the number of vulnerabilities reported. Do you think something like that would be useful?

arcanis commented:
> But I was just going to say having the whole paths tree might be too much, but what do you think about just keeping the top-level (direct dependencies) in the report?

That seems reasonable 🤔 Perhaps vertical space can also be gained by printing the direct dependency names on a single line, comma-separated?

> So the equivalent of `npm audit` would be `yarn npm audit --recursive` (and yarn classic equivalent of `yarn audit` would be `yarn npm audit --all --recursive`) but now we have the added functionality of being able to limit that audit report to just direct dependencies?

Yep!

> I think in my last commit I had a bit of a half-baked idea of just printing the number of vulnerabilities reported. Do you think something like that would be useful?

Oh right - you were printing the numbers as returned from the registry, in a format similar to `npm audit`. I tried removing it since those numbers often feel off and don't provide much value - I suspect it's intended more as a marketing ploy than anything, and I'm not a fan of that ("look how many packages we've analyzed for you!").

If we add it back, I suspect the only number that matters is the number of reports, and it's awkward to integrate with the tree display, so I'd refrain from adding it for now and see if it's actually a problem depending on the usage.

jdanil commented:
> That seems reasonable 🤔 Perhaps vertical space can also be gained by printing the direct dependency names on a single line, comma-separated?
> 
> >

Done! Feel free to update it, or remove it if you don't like it. I updated the docs to include more information on how to determine the dependency tree for the vulnerable packages, including using `yarn why`.
Do you think it would make sense to add something like a `--verbose` flag that prints out the full path tree too? I'm thinking there may be cases where it would still use useful to see that whole dependency tree, as `yarn why` is a bit of a flat list. It may be useful for vulnerabilities that don't have 100s of packages depending on them.

> If we add it back, I suspect the only number that matters is the number of reports, and it's awkward to integrate with the tree display, so I'd refrain from adding it for now and see if it's actually a problem depending on the usage.

Yep no worries. I was adding it more for consistency with npm, yarn classic, and snyk than anything else. Don't think its all that important. I think people would care more about either the JSON output or the exit code.

I also updated the output tree to respect the severity flag, so it will only print vulnerabilities over that threshold.

The last commit I pushed up includes a fix for the `400 (Bad Request)` error I was getting when running `yarn npm audit` without `--recursive` (I wonder if you were getting it too). I believe this request was returning an error because the npm endpoint knew the dependency list was not complete by looking at the required dependencies for each package.

jdanil commented:
@arcanis let me know if you're waiting on anything else from me on this.

arcanis commented:
All good, thanks @jdanil! 🎉

jdanil commented:
No worries. Thanks for your help @arcanis and @lirantal 🙏

lirantal commented:
Great job @jdanil 👏   👏 👏

simonespa commented:
Thanks guys for this effort 👏 👏
